### PR TITLE
Allow disruptive conformance tests and tag 2 as such

### LINKED ISSUE #82663: Some taint tests will evict non-test workloads as a result and missing [Disruptive]
**What happened**:
Ran a conformance test run (using Sonobuoy) on v1.16.0.beta.3

One of the taint based tests added a NoExecute taint to a node which evicted all workloads including the Sonobuoy workload itself. If this had been a prod workload, numerous prod workloads could have been impacted.

**What you expected to happen**:
Test run as normal with reasonable results. Able to run conformance tests with minimal impact on a running cluster.

**How to reproduce it (as minimally and precisely as possible)**:
`sonobuoy run` on a v1.16.0 cluster.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.16.0.beta.3

/area conformance

### LINKED ISSUE #82787: Allow conformance tests to be marked as Disruptive
**What would you like to be added**:
Our tooling that makes the list of conformance tests needs to allow conformance tests to be marked as disruptive.

**Why is this needed**:
To test default, conformance behavior which may be disruptive (e.g. taint eviction)

ref #82663

/sig testing
/area conformance

### PR DESCRIPTION
**Pull request by johnSchnake**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
- Adjusts tooling to allow disrutpive tests to be marked conformance
 - Adds the disruptive tag to both tests (disruptive due to tainting
nodes and evicting other workloads)

**Which issue(s) this PR fixes**:
Fixes #82663
Fixes #82787

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Conformance tests may now include disruptive tests. If you are running tests against a live cluster, consider skipping those tests tagged as `Disruptive` to avoid non-test workloads being impacted. Be aware, skipping any conformance tests (even disruptive ones) will make the results ineligible for consideration for the CNCF Certified Kubernetes program.
```

### CODE REVIEW

File: test/e2e/scheduling/taints.go [hunk 1]
```diff
@@ -283,7 +283,7 @@ var _ = SIGDescribe("NoExecuteTaintManager Single Pod [Serial]", func() {
 		Description: The Pod with toleration timeout scheduled on a tainted Node MUST not be
 		evicted if the taint is removed before toleration time ends.
 	*/
-	framework.ConformanceIt("removing taint cancels eviction", func() {
+	framework.ConformanceIt("removing taint cancels eviction [Disruptive]", func() {
```
[the remark this answers is not in the document]
Comment (johnSchnake) [reply test/e2e/scheduling/taints.go]: I am definitely not versed in how the scheduling/eviction works, but I'm just confused because your description doesn't seem to square with what is happening.

I didn't trust my understanding so I wanted to repro it myself. I built the kind images for the v1.16.0-rc.2 tag and ran just the `removing taint cancels eviction` test and reprod that the sonobuoy pod (a workload outside the test framework) gets evicted.

Here are some of the logs from the controller:
```
0913 18:54:07.405937       1 taint_manager.go:105] NoExecuteTaintManager is deleting Pod: heptio-sonobuoy/sonobuoy-e2e-job-3223075635be415d
I0913 18:54:07.406113       1 taint_manager.go:105] NoExecuteTaintManager is deleting Pod: heptio-sonobuoy/sonobuoy
I0913 18:54:07.406147       1 taint_manager.go:105] NoExecuteTaintManager is deleting Pod: kube-system/kindnet-lpx4w
I0913 18:54:07.406675       1 event.go:255] Event(v1.ObjectReference{Kind:"Pod", Namespace:"heptio-sonobuoy", Name:"sonobuoy", UID:"", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'TaintManagerEviction' Marking for deletion Pod heptio-sonobuoy/sonobuoy
I0913 18:54:07.406727       1 event.go:255] Event(v1.ObjectReference{Kind:"Pod", Namespace:"heptio-sonobuoy", Name:"sonobuoy-e2e-job-3223075635be415d", UID:"", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'TaintManagerEviction' Marking for deletion Pod heptio-sonobuoy/sonobuoy-e2e-job-3223075635be415d
```

Is this a bug? I see the eviction/deletion of the pod happen even before it logs that it is marking it for eviction.

[the remark this answers is not in the document]
Comment (johnSchnake) [reply test/e2e/scheduling/taints.go]: I think everyone here would welcome a PR if it made the tags more appropriate on the rest of the tests.  These two, being conformance related, have just a higher priority so I want to be sure to handle them for the release.

**Comment by johnSchnake** [comment]
@neolit123 @timothysc @johnbelamaric So the `walk_test.go` file actually automated the fact that `[Disruptive]` tests cant be conformance and so the automated script to generate the conformance list won't recognize these tests.

Should I, in this PR, also be making disruptive tests allowed via that code?

It just seems like we are very close to the v1.16 release and we don't have clear agreement about whether or not this test should be disruptive, why it is disruptive, or if disruptive tests should be conformance tests.

Unsure how to proceed.

**Comment by neolit123** [comment]
> So the walk_test.go file actually automated the fact that [Disruptive] tests cant be conformance and so the automated script to generate the conformance list won't recognize these tests.

no, that was the whole point to allow the limitation. @spiffxp has more context.
if a test is disruptive it's not conformance and your observation proves that.

> It just seems like we are very close to the v1.16 release and we don't have clear agreement about whether or not this test should be disruptive, why it is disruptive, or if disruptive tests should be conformance tests.

i'm bringing this to sig-release.

**Comment by dims** [comment]
@johnSchnake @neolit123 is there a changet we can make to the test for it to avoid the scenario where the pod that is running the e2e conformance tests do not get nuked? (like look up the "current" node using downard api and avoid scheduling the test pod in the current node?)

**Comment by johnSchnake** [comment]
>  is there a change we can make to the test for it to avoid the scenario where the pod that is running the e2e conformance tests do not get nuked? 

I don't think that approach is terribly useful; it avoids one issue in some situations:
 - What if it is a 2 node cluster (1 master 1 worker); are those no longer capable of conformance? Or they'll still just hit this issue anyways? Do we fail since we can't avoid restarting the e2e pod? Do we warn and do it anyways?
 - 1 node clusters couldn't pass conformance anyways, but it would still cause the e2e pod to get evicted anyways (or fail/warn like above)
 - What about other workloads besides the e2e pod? (prod workloads, sonobuoy workload, etc).

I've already committed a PR to Sonobuoy so that we tolerate this taint and launch the e2e plugin to tolerate this taint. It is a bit of a stopgap but I feel like if we are going to allow 'destructive' tests then it makes sense to at least clearly document/limit the ways in which they intend to be destructive so that users can mitigate any risks.

**Comment by dims** [comment]
@johnSchnake i guess then decision is made already based on @bgrant0607 and @johnbelamaric 's notes above. Also the existing rule we have that we should *not* have any tags for conformance tests. Only thing left now is to document what was said earlier.

**Comment by johnSchnake** [comment]
Also, on the topic of what conformance is for, @bgrant0607 gave a link to https://www.cncf.io/certification/software-conformance/ but despite describing all the benefits for being run for providers etc etc it also says:

> Confirmability: any end user can confirm that their distribution or platform remains conformant by running the identical open source conformance application (Sonobuoy) that was used to certify.

Which seems to indicate an intent/capability to run them on a prod cluster.

**Comment by dims** [comment]
@johnSchnake let's fix that doc to reflect the current thoughts.

**Comment by johnSchnake** [comment]
We have 2 days before the release and currently, by default, run secretly disruptive tests in a conformance run.

IMO that is the worst possible case; the alternatives:
1. demote the 2 tests (temporarily maybe) and resolve this after v1.16
2.  allow destructive+conformance tests
    - update the scripts that generate the conformance list to allow this _NEW_ behavior
    - update the cncf documentation about how to get certification
    - update Sonobuoy to support this appropriately/safely
    - add the tags to the test

Due to the larger scope of (2) and the short time horizon, I advocate for (1). Even if that means we pursue (2) right after the release, it seems too rushed a decision to make on the deadline we have.

Thoughts?

**Comment by johnSchnake** [comment]
See updated ticket with information about resolution. TL;DR; we are leaving tests as-is (mis-tagged) for v1.16 and dealing with it.

This PR is now targeted for v1.17 and includes the changes to support disruptive conformance tests based on the conversations I've had about the issue.

**Comment by johnSchnake** [comment]
/hold

Waiting for confirmation during a conformance-wg meeting that this is 100% the direction to go. Seems like we have a fair amount of consensus on slack though.

**Comment by johnbelamaric** [comment]
FYI it would be the conformance subproject of SIG-arch that decides, not the CNCF conformance-wg.

I think that's what you meant anyway, but want to make it clear. I see you already added it to the [agenda](https://docs.google.com/document/d/1W31nXh9RYAb_VaYkwuPLd1hFxuRX3iU0DmaQ4lkCsX8/edit#heading=h.l87lu17xm9bh). Thanks!

File: test/e2e/scheduling/taints.go (hunk 1 above)
Comment (johnbelamaric) [inline test/e2e/scheduling/taints.go]: Please split changes to this file into a separate PR.

Comment (timothysc) [reply test/e2e/scheduling/taints.go]: I kind of think this is fine right now and shouldn't affect the suite.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 
/hold 
^ We can talk about this next Tuesday.

File: test/e2e/scheduling/taints.go (hunk 1 above)
Comment (johnbelamaric) [reply test/e2e/scheduling/taints.go]: /shrug

Comment (johnSchnake) [reply test/e2e/scheduling/taints.go]: OK, I rebased onto master. Left this in the PR for now unless the request is reinforced.

**Comment by johnSchnake** [comment]
This was discussed in the conformance-wg meeting; there was agreement that this is appropriate (to have disruptive conformance tests) so I'm removing the hold.

/hold cancel

**Comment by dims** [comment]
/lgtm

(re-applying after rebase)

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve
